src/server/cluster/cluster_family.cc

```diff
@@ -0,0 +1,266 @@
+                                   epoch_master_time, " 1 ", connect_state, "\r\n");
+    absl::StrAppend(&msg, server_family_->GetReplicaMasterId(), " ", info.host, ":", info.port, "@",
+                    info.port, " master - 0 ", epoch_master_time, " 1 ", connect_state,
+                    " 0-16383\r\n");
+    return msg;
+  }
+}
+
+void ClusterFamily::ClusterHelp(ConnectionContext* cntx) {
+  string_view help_arr[] = {
+      "CLUSTER <subcommand> [<arg> [value] [opt] ...]. Subcommands are:",
+      "SLOTS",
+      "   Return information about slots range mappings. Each range is made of:",
+      "   start, end, master and replicas IP addresses, ports and ids.",
+      "NODES",
+      "   Return cluster configuration seen by node. Output format:",
+      "   <id> <ip:port> <flags> <master> <pings> <pongs> <epoch> <link> <slot> ...",
+      "INFO",
+      "  Return information about the cluster",
+      "HELP",
+      "    Prints this help.",
+  };
+  return (*cntx)->SendSimpleStrArr(help_arr);
+}
+
+void ClusterFamily::ClusterSlots(ConnectionContext* cntx) {
+  // For more details https://redis.io/commands/cluster-slots/
+  constexpr unsigned int kClustersShardingCount = 1;
+  constexpr unsigned int kNoReplicaInfoSize = 3;
+  constexpr unsigned int kWithReplicaInfoSize = 4;
+
+  /* Format: 1) 1) start slot
+   *            2) end slot
+   *            3) 1) master IP
+   *               2) master port
+   *               3) node ID
+   *            4) 1) replica IP (optional)
+   *               2) replica port
+   *               3) node ID
+   *           ... note that in this case, only 1 slot
+   */
+  ServerState& etl = *ServerState::tlocal();
+  // we have 3 cases here
+  // 1. This is a stand alone, in this case we only sending local information
+  // 2. We are the master, and we have replica, in this case send us as master
+  // 3. We are replica to a master, sends the information about us as replica
+  (*cntx)->StartArray(kClustersShardingCount);
+  if (etl.is_master) {
+    std::string cluster_announce_ip = absl::GetFlag(FLAGS_cluster_announce_ip);
+    std::string preferred_endpoint =
+        cluster_announce_ip.empty() ? cntx->owner()->LocalBindAddress() : cluster_announce_ip;
+    auto vec = server_family_->GetDflyCmd()->GetReplicasRoleInfo();
+    unsigned int info_len = vec.empty() ? kNoReplicaInfoSize : kWithReplicaInfoSize;
+    (*cntx)->StartArray(info_len);
+    (*cntx)->SendLong(0);                           // start sharding range
+    (*cntx)->SendLong(ClusterConfig::kMaxSlotNum);  // end sharding range
+    BuildClusterSlotNetworkInfo(cntx, preferred_endpoint, absl::GetFlag(FLAGS_port),
+                                server_family_->master_id());
+    if (!vec.empty()) {  // info about the replica
+      const auto& info = vec[0];
+      BuildClusterSlotNetworkInfo(cntx, info.address, info.listening_port, etl.remote_client_id_);
+    }
+  } else {
+    Replica::Info info = server_family_->GetReplicaInfo();
```

Comment:
There is a bug here that if somesone run replicaof no one while this command is running you can get a segfault here as GetReplicaInfo accesses the replica_ ptr without checking its not null.
I believe this bug was also before this PR. 
Please fix this in another PR, not this one

Reply:
Sounds like a job for `MutexGuardedObject`! :)
As you said, that can be in another PR though. And there might be a simpler way too :)